server/src/test/java/org/apache/druid/segment/loading/SegmentLoaderLocalCacheManagerTest.java

```diff
@@ -372,24 +372,276 @@ public void testEmptyToFullOrder() throws Exception
   }
 
   private DataSegment dataSegmentWithInterval(String intervalStr)
+  {
+    return dataSegmentWithInterval(intervalStr, 10L);
+  }
+
+  private DataSegment dataSegmentWithInterval(String intervalStr, long size)
   {
     return DataSegment.builder()
-                      .dataSource("test_segment_loader")
-                      .interval(Intervals.of(intervalStr))
-                      .loadSpec(
-                          ImmutableMap.of(
-                              "type",
-                              "local",
-                              "path",
-                              "somewhere"
-                          )
-                      )
-                      .version("2015-05-27T03:38:35.683Z")
-                      .dimensions(ImmutableList.of())
-                      .metrics(ImmutableList.of())
-                      .shardSpec(NoneShardSpec.instance())
-                      .binaryVersion(9)
-                      .size(10L)
-                      .build();
+      .dataSource("test_segment_loader")
+      .interval(Intervals.of(intervalStr))
+      .loadSpec(
+        ImmutableMap.of(
+          "type",
+          "local",
+          "path",
+          "somewhere"
+        )
+      )
+      .version("2015-05-27T03:38:35.683Z")
+      .dimensions(ImmutableList.of())
+      .metrics(ImmutableList.of())
+      .shardSpec(NoneShardSpec.instance())
+      .binaryVersion(9)
+      .size(size)
+      .build();
+  }
+
+  @Test
+  public void testSegmentDistributionUsingRoundRobinStrategy() throws Exception
```

Comment:
I suggest to add a new class to test `StorageLocationSelectorStrategy` instead these tests. The new tests here are quite duplicate and incomplete. The unit tests should verify that the locations returned by the strategy is ordered correctly in various situations including when multiple threads are calling it.

Reply:
> The new tests here are quite duplicate and incomplete. The unit tests should verify that the locations returned by the strategy is ordered correctly in various situations

1. The tests do verify the locations returned by the strategy are ordered correctly or not. 
Common reusable bits of code have been extracted into methods.

In `testSegmentDistributionUsingRoundRobinStrategy` . 
The below comments in the test make it clear that for round robin strategy segment1 is downloaded into local_storage_folder, segment2 to folder2, segment3 to folder3 and that the segment4 is downloaded into local_storage_folder asserting round robin behavior.  
```
// Segment 1 should be downloaded in local_storage_folder
// Segment 2 should be downloaded in local_storage_folder2
// Segment 3 should be downloaded in local_storage_folder3
// Segment 4 should be downloaded in local_storage_folder again, asserting round robin distribution of segments
```
One case I could add was setting a `localStorageFolder` unwritable, but this is covered in tests `testRetrySuccessAtSecondLocation` and others. Moreover writable or not isn't checked by the strategy. Let me know if I miss any other cases.

2. Changes in existing tests cover `LeastBytesUsedStorageLocationSelectorStrategy` as this is the default strategy.

>  including when multiple threads are calling it.
This is something I missed. Is `SegmentLoaderLocalCacheManagerConcurrencyTest` a write place to add ?

Reply:
>  including when multiple threads are calling it.

This is something I missed. Is `SegmentLoaderLocalCacheManagerConcurrencyTest` a write place to add ?

Reply:
What I'm saying is this is not a good place to test `StorageLocationSelectorStrategy`. The newly added tests here test some part of `StorageLocationSelectorStrategy` but are not fully testing in various situations. Also, they are duplicate with other tests in this class because they should end up calling `SegmentLoaderLocalCacheManager` API. I think most of the codes in the new tests are not necessary if we add a new test class for testing only `StorageLocationSelectorStrategy`. 

Similarly, `SegmentLoaderLocalCacheManagerConcurrencyTest` is to do some concurrency test for `SegmentLoaderLocalCacheManager` not `StorageLocationSelectorStrategy`. Please add a new class for it.

Reply:
Got it. I'll add new test classes to test `StorageLocationSelectorStrategy` and concurrency for it.

Reply:
Added a couple of tests in StorageLocationSelectorStrategyTest.